documentation/design-docs/ipc-protocol.md

```diff
@@ -0,0 +1,441 @@
+    <th>20</th>
+    <th>21</th>
+    <th>22</th>
+    <th>23</th>
+    <th>24</th>
+    <th>...</th>
+    <th>size - 1 </th>
+    <th>size</th>
+  </tr>
+  <tr>
+    <td colspan="20">header</td>
+    <td colspan="7">payload</td>
+  </tr>
+  <tr>
+    <td colspan="14">magic</td>
+    <td colspan="2">size</td>
+    <td colspan="1">command_set</td>
+    <td colspan="1">command_id</td>
+    <td colspan="2">reserved</td>
+    <td colspan="7">payload</td>
+  </tr>
+</table>
+
+The simplest Diagnostic IPC Message will contain a Header and an empty Payload and therefore only be 20 bytes long.
+
+For example, this IPC Message sends a hypothetical Command (discussed later) `0xFFEE` which has an empty Payload:
+<table>
+  <tr>
+    <th>1</th>
+    <th>2</th>
+    <th>3</th>
+    <th>4</th>
+    <th>5</th>
+    <th>6</th>
+    <th>7</th>
+    <th>8</th>
+    <th>9</th>
+    <th>10</th>
+    <th>11</th>
+    <th>12</th>
+    <th>13</th>
+    <th>14</th>
+    <th>15</th>
+    <th>16</th>
+    <th>17</th>
+    <th>18</th>
+    <th>19</th>
+    <th>20</th>
+  </tr>
+  <tr>
+    <tr>
+    <td colspan="14">magic</td>
+    <td colspan="2">size</td>
+    <td colspan="2">command</td>
+    <td colspan="2">reserved</td>
+  </tr>
+  <tr>
+    <td colspan="14">"DOTNET_IPC_V1"</td>
+    <td colspan="2">20</td>
+    <td colspan="1">0xFF</td>
+    <td colspan="1">0xEE</td>
+    <td colspan="2">0x0000</td>
+  </tr>
+</table>
+
+#### Headers
+
+Every Diagnostic IPC Message will start with a header and every header will:
+* start with a magic version number and a size
+* `sizeof(IpcHeader) == 20`
+* encode numbers little-endian
+* account for the size of the payload in the `size` value, i.e., `IpcHeader.size == sizeof(IpcHeader) + PayloadStruct.GetSize()`
+
+```c
+// size = 14 + 2 + 1 + 1 + 2 = 20 bytes
+struct IpcHeader
+{
+    char[14]  magic;    // null terminated string of the version of the protocol, for example "DOTNET_IPC_V1"
+    uint16_t  size;     // size of packet = size of header + payload
+    uint8_t   command_set;  // combined with command_id is the Command to invoke
+    uint8_t   command_id;   // combined with command_set is the Command to invoke
+    uint16_t  reserved; // for potential future use
+};
+```
+
+The `reserved` field is reserved for future use.  It is unused in `DOTNET_IPC_V1`, so it should be set to `0x0000` by convention when unused.
+
+
+#### Payloads
+
+Payloads are Command specific data encoded into a Diagnostic IPC Message.  The size of the payload is implicitly encoded in the Header's `size` field as `PayloadSize = header.size - sizeof(struct IpcHeader)`.  A Payload _may_ be 0 bytes long if it empty.  The encoding of data in the Payload is Command specific.
+
+As an example, EventPipe encodes non fixed-size Payloads using type codes, little-endian numbers, and size-prefixed, null-terminated char strings.  Using these rules, a hypothetical Diagnostic IPC Message telling EventPipe to start streaming _may_ look like the following:
```

Comment:
size-prefixed -> you will need to specify the size of the size prefix (ie uint8, uint16, uint32?)
char string -> 'char' is a bit ambiguous in some contexts it is one byte, in other places it is 2 or 4 bytes. We'll need to nail this down.
For the strings we should specify the encoding. I think we are currently using UTF16 and I recommend we stick with it for simplicity. This also means we have to specify whether the size prefix is a count of bytes or a count of characters, and whether or not the null-terminator is included in the count.